llvm/lib/Analysis/DXILMetadataAnalysis.cpp

```diff
@@ -0,0 +1,151 @@
+}
+
+void ModuleMetadataInfo::init(Module &M) {
+  Triple TT(Triple(M.getTargetTriple()));
+  DXILVersion = TT.getDXILVersion();
+  ShaderModelVersion = TT.getOSVersion();
+  ShaderStage = TT.getEnvironment();
+  NamedMDNode *Entry = M.getOrInsertNamedMetadata("dx.valver");
+  if (Entry->getNumOperands() == 0) {
+    ValidatorVersion = VersionTuple(1, 0);
+  } else {
+    auto *ValVerMD = cast<MDNode>(Entry->getOperand(0));
+    auto *MajorMD = mdconst::extract<ConstantInt>(ValVerMD->getOperand(0));
+    auto *MinorMD = mdconst::extract<ConstantInt>(ValVerMD->getOperand(1));
+    ValidatorVersion =
+        VersionTuple(MajorMD->getZExtValue(), MinorMD->getZExtValue());
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// DXILMetadataAnalysis and DXILMetadataAnalysisPrinterPass
+
+// Provide an explicit template instantiation for the static ID.
+AnalysisKey DXILMetadataAnalysis::Key;
+
+llvm::dxil::ModuleMetadataInfo
+DXILMetadataAnalysis::run(Module &M, ModuleAnalysisManager &AM) {
+  Data.init(M);
+  return Data;
+}
+
+PreservedAnalyses
+DXILMetadataAnalysisPrinterPass::run(Module &M, ModuleAnalysisManager &AM) {
+  llvm::dxil::ModuleMetadataInfo &Data = AM.getResult<DXILMetadataAnalysis>(M);
+
+  Data.dump(OS);
+  return PreservedAnalyses::all();
+}
+
+//===----------------------------------------------------------------------===//
+// DXILMetadataAnalysisLegacyPass
+
+DXILMetadataAnalysisLegacyPass::DXILMetadataAnalysisLegacyPass()
+    : ModulePass(ID) {
+  initializeDXILMetadataAnalysisLegacyPassPass(
+      *PassRegistry::getPassRegistry());
+}
+
+DXILMetadataAnalysisLegacyPass::~DXILMetadataAnalysisLegacyPass() = default;
+
+void DXILMetadataAnalysisLegacyPass::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.setPreservesAll();
+}
+
+bool DXILMetadataAnalysisLegacyPass::runOnModule(Module &M) {
+  Data.init(M);
+  AnalysisDone = true;
+  return false;
+}
+
+void DXILMetadataAnalysisLegacyPass::print(raw_ostream &OS,
+                                           const Module *) const {
+  if (!AnalysisDone) {
+    OS << "No module metadata info has been built!\n";
+    return;
+  }
+  Data.dump();
+}
+
+#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
+LLVM_DUMP_METHOD
+void DXILMetadataAnalysisLegacyPass::dump() const { print(dbgs(), nullptr); }
+#endif
+
+INITIALIZE_PASS(DXILMetadataAnalysisLegacyPass, "dxil-metadata-analysis",
+                "DXIL Module Metadata analysis", false, true)
+char DXILMetadataAnalysisLegacyPass::ID = 0;
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Pass to print - for verification - of analysis information collected in
+// DXILMetadataAnalysisLegacyPass
+
+class DXILMetadataAnalysisLegacyPrinter : public ModulePass {
+public:
+  static char ID;
+
+  DXILMetadataAnalysisLegacyPrinter();
+
+  bool runOnModule(Module &M) override;
+  void getAnalysisUsage(AnalysisUsage &AU) const override;
+};
```

Comment:
I don't know that we need to bother with the legacy printer pass - we can just use the NewPM printer pass in the tests (that is, `opt -passes=print<dxil-metadata>` instead of `-dxil-metadata-analysis-print`)